File: app/src/main/java/com/example/bookapp/MyApplication.java
package com.example.bookapp;

import static com.example.bookapp.Constants.MAX_BYTES_PDF;

import android.app.Application;
import android.app.ProgressDialog;
import android.content.Context;
import android.os.Environment;
import android.text.format.DateFormat;
import android.util.Log;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.example.bookapp.adapter.AdapterPdfAdmin;
import com.example.bookapp.model.ModelPdf;
import com.github.barteksc.pdfviewer.PDFView;
import com.github.barteksc.pdfviewer.listener.OnErrorListener;
import com.github.barteksc.pdfviewer.listener.OnLoadCompleteListener;
import com.github.barteksc.pdfviewer.listener.OnPageErrorListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageMetadata;
import com.google.firebase.storage.StorageReference;

import java.io.File;
import java.io.FileOutputStream;
import java.util.Calendar;
import java.util.HashMap;
import java.util.Locale;

public class MyApplication extends Application {

    public static final String TAG_DOWN = "DOWN_BOOK";


    @Override
    public void onCreate() {
        super.onCreate();
    }

    public static String formatTimeStamp(long timestamp){
        Calendar cal = Calendar.getInstance(Locale.ENGLISH);
        cal.setTimeInMillis(timestamp);

        String date = DateFormat.format("dd/mm/yyyy", cal).toString();

        return date;

    }

    public static void deleteBook(Context context, String bookId, String bookUrl, String bookTitle) {
        String TAG = "DELETE_BOOK";

        Log.d(TAG, "deleteBook: ");

        ProgressDialog progressDialog = new ProgressDialog(context);
        progressDialog.setTitle("Please wait");
        progressDialog.setMessage("Deleting" +bookTitle+"..." );
        progressDialog.show();

        //xoa trong storage neu thanh cong thi xoa trong db
        StorageReference storageReference = FirebaseStorage.getInstance().getReferenceFromUrl(bookUrl);
        storageReference.delete()
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void unused) {
                        DatabaseReference reference = FirebaseDatabase.getInstance().getReference("Books");
                        reference.child(bookId)
                                .removeValue()
                                .addOnSuccessListener(new OnSuccessListener<Void>() {
                                    @Override
                                    public void onSuccess(Void unused) {
                                        progressDialog.dismiss();
                                        Toast.makeText(context, "Book delete successfully", Toast.LENGTH_SHORT).show();
                                    }
                                })
                                .addOnFailureListener(new OnFailureListener() {
                                    @Override
                                    public void onFailure(@NonNull Exception e) {
                                        progressDialog.dismiss();
                                        Toast.makeText(context, ""+e.getMessage(), Toast.LENGTH_SHORT).show();
                                    }
                                });
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        progressDialog.dismiss();
                        Toast.makeText(context, ""+e.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
    }

    public static void loadPdfSize(String pdfUrl, String pdfTitle, TextView sizeTv) {
        String TAG = "BOOK_SIZE_TAG";


        StorageReference ref = FirebaseStorage.getInstance().getReferenceFromUrl(pdfUrl);
        ref.getMetadata()
                .addOnSuccessListener(new OnSuccessListener<StorageMetadata>() {
                    @Override
                    public void onSuccess(StorageMetadata storageMetadata) {
                        Log.d(TAG, "onSuccess: "+ pdfTitle);

                        double bytes = storageMetadata.getSizeBytes();

                        double kb = bytes/1024;
                        double mb = kb/1024;
                        if(mb >= 1){
                            sizeTv.setText(String.format("%.2f",mb) + " MB");
                        }
                        else if(kb >= 1) {
                            sizeTv.setText(String.format("%.2f",kb) + " KB");
                        }
                        else {
                            sizeTv.setText(String.format("%.2f",bytes) + " bytes");
                        }

                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Log.d(TAG, "onFailure: "+e.getMessage());
                    }
                });
    }

    public static void loadBannerPdf(String pdfUrl, String pdfTitle, PDFView pdfView, ProgressBar progressBar, TextView pageTv) {

        String TAG = "LOAD_BANNER_PDF";

        Log.d(TAG, "loadBannerPdf: url la: "+ pdfUrl);

        StorageReference ref = FirebaseStorage.getInstance().getReferenceFromUrl(pdfUrl);
        ref.getBytes(MAX_BYTES_PDF)
                .addOnSuccessListener(new OnSuccessListener<byte[]>() {
                    @Override
                    public void onSuccess(byte[] bytes) {
                        Log.d(TAG, "onSuccess: load"+ pdfTitle);

                        pdfView.fromBytes(bytes)
                                .pages(0)// show ra trang dau tien cua pdf
                                .spacing(0)
                                .swipeHorizontal(false)
                                .enableSwipe(false)
                                .onError(new OnErrorListener() {
                                    @Override
                                    public void onError(Throwable t) {
                                        progressBar.setVisibility(View.INVISIBLE);
                                        Log.d(TAG, "onError: "+ t.getMessage());
                                    }
                                })
                                .onPageError(new OnPageErrorListener() {
                                    @Override
                                    public void onPageError(int page, Throwable t) {
                                        progressBar.setVisibility(View.INVISIBLE);
                                        Log.d(TAG, "onPageError: "+ t.getMessage());
                                    }
                                })
                                .onLoad(new OnLoadCompleteListener() {
                                    @Override
                                    public void loadComplete(int nbPages) {
                                        // tai trang pdf thi se an cai xoay xoay
                                        progressBar.setVisibility(View.INVISIBLE);

                                        if(pageTv != null){
                                            pageTv.setText(""+nbPages);
                                        }
                                    }
                                })
                                .load();


                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Log.d(TAG, "onFailure: "+e.getMessage());
                    }
                });

    }

    public static void loadCategory(String categoryId, TextView categoryTv) {


        DatabaseReference ref = FirebaseDatabase.getInstance().getReference("Categories");
        ref.child(categoryId)
                .addListenerForSingleValueEvent(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot snapshot) {
                        String category = ""+snapshot.child("category").getValue();

                        categoryTv.setText(category);
                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError error) {

                    }
                });
    }

    public static void incrementBookCount(String bookId){
        DatabaseReference ref = FirebaseDatabase.getInstance().getReference("Books");
        ref.child(bookId)
                .addListenerForSingleValueEvent(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot snapshot) {
                        String viewCount = ""+snapshot.child("viewCount").getValue();
                        if(viewCount.equals("") || viewCount.equals("null")){
                            viewCount = "0";
                        }

                        long newViewCount = Long.parseLong(viewCount) + 1;

                        HashMap<String, Object> hashMap = new HashMap<>();
                        hashMap.put("viewCount", newViewCount);

                        DatabaseReference reference = FirebaseDatabase.getInstance().getReference("Books");
                        reference.child(bookId)
                                .updateChildren(hashMap);

                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError error) {

                    }
                });
    }

    public static void downloadBook(Context context, String bookId, String bookTitle, String bookUrl){

        String namepdf = bookTitle + ".pdf";
        Log.d(TAG_DOWN, "downloadBook: "+namepdf);

        ProgressDialog progressDialog = new ProgressDialog(context);
        progressDialog.setTitle("Please wait");
        progressDialog.setMessage("Downloading "+ namepdf);
        progressDialog.setCanceledOnTouchOutside(false);
        progressDialog.show();

        StorageReference storageReference = FirebaseStorage.getInstance().getReferenceFromUrl(bookUrl);
        storageReference.getBytes(MAX_BYTES_PDF)
                .addOnSuccessListener(new OnSuccessListener<byte[]>() {
                    @Override
                    public void onSuccess(byte[] bytes) {
//                        progressDialog.dismiss();
                        Log.d(TAG_DOWN, "onSuccess: Saving book");
                        saveDownloadBook(context, progressDialog, bytes, namepdf, bookId);
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        progressDialog.dismiss();
                        Toast.makeText(context, "Failed to download "+e.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });

    }

    private static void saveDownloadBook(Context context, ProgressDialog progressDialog, byte[] bytes, String namepdf, String bookId) {
        Log.d(TAG_DOWN, "saveDownloadBook: Saving download book");
        try{
            File downloadFolder  = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);
            downloadFolder.mkdirs();

            String filePath = downloadFolder.getPath() + "/" + namepdf;

            FileOutputStream out = new FileOutputStream(filePath);
            out.write(bytes);
            out.close();

            Toast.makeText(context, "Saved to Download Folder", Toast.LENGTH_SHORT).show();
            progressDialog.dismiss();


            // tang luot download len
            incrementBookDownloadCount(bookId);

        }catch (Exception e){
            Log.d(TAG_DOWN, "saveDownloadBook: Failed saving to Download Folder: "+e.getMessage());
            Toast.makeText(context, "Failed saving to Download: " + e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    private static void incrementBookDownloadCount(String bookId) {
        DatabaseReference ref = FirebaseDatabase.getInstance().getReference("Books");
        ref.child(bookId)
                .addListenerForSingleValueEvent(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot snapshot) {
                        String downloadCount = ""+snapshot.child("downloadCount").getValue();
                        Log.d(TAG_DOWN, "onDataChange: DL count: " + downloadCount);
                        if(downloadCount.equals("") || downloadCount.equals("null")){
                            downloadCount = "0";
                        }

                        long newDownloadCount = Long.parseLong(downloadCount) + 1;

                        HashMap<String, Object> hashMap = new HashMap<>();
                        hashMap.put("downloadCount", newDownloadCount);

                        DatabaseReference reference = FirebaseDatabase.getInstance().getReference("Books");
                        reference.child(bookId)
                                .updateChildren(hashMap)
                                .addOnSuccessListener(new OnSuccessListener<Void>() {
                                    @Override
                                    public void onSuccess(Void unused) {
                                        Log.d(TAG_DOWN, "onSuccess: incrementBookDownload ");
                                    }
                                })
                                .addOnFailureListener(new OnFailureListener() {
                                    @Override
                                    public void onFailure(@NonNull Exception e) {
                                        Log.d(TAG_DOWN, "onFailure: failed update "+ e.getMessage());
                                    }
                                });

                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError error) {

                    }
                });


    }

    public static void addToFavorite(Context context, String bookId){
        // dang nhap moi them duoc
        FirebaseAuth firebaseAuth = FirebaseAuth.getInstance();
        if(firebaseAuth.getCurrentUser() == null){
            Toast.makeText(context, "You are not logged in", Toast.LENGTH_SHORT).show();
        }
        else {
            long timestamp = System.currentTimeMillis();

            HashMap<String, Object> hashMap = new HashMap<>();
            hashMap.put("bookId", ""+bookId);
            hashMap.put("timestamp", ""+timestamp);

            DatabaseReference ref = FirebaseDatabase.getInstance().getReference("Users");
            ref.child(firebaseAuth.getUid()).child("Favorites").child(bookId)
                    .setValue(hashMap)
                    .addOnSuccessListener(new OnSuccessListener<Void>() {
                        @Override
                        public void onSuccess(Void unused) {
                            Toast.makeText(context, "Added to your favotite list", Toast.LENGTH_SHORT).show();
                        }
                    })
                    .addOnFailureListener(new OnFailureListener() {
                        @Override
                        public void onFailure(@NonNull Exception e) {
                            Toast.makeText(context, "Failed add to your favotite list "+e.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    });

        }

    }

    public static void removeFromFavorite(Context context, String bookId){
        // dang nhap moi them duoc
        FirebaseAuth firebaseAuth = FirebaseAuth.getInstance();
        if(firebaseAuth.getCurrentUser() == null){
            Toast.makeText(context, "You are not logged in", Toast.LENGTH_SHORT).show();
        }
        else {
            DatabaseReference ref = FirebaseDatabase.getInstance().getReference("Users");
            ref.child(firebaseAuth.getUid()).child("Favorites").child(bookId)
                    .removeValue()
                    .addOnSuccessListener(new OnSuccessListener<Void>() {
                        @Override
                        public void onSuccess(Void unused) {
                            Toast.makeText(context, "Removed to your favotite list", Toast.LENGTH_SHORT).show();
                        }
                    })
                    .addOnFailureListener(new OnFailureListener() {
                        @Override
                        public void onFailure(@NonNull Exception e) {
                            Toast.makeText(context, "Failed to remove to your favotite list "+e.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    });

        }
    }

}

File: app/src/main/java/com/pluto/bookapp/MyApplication.kt
package com.pluto.bookapp

import android.app.Application
import android.app.ProgressDialog
import android.content.Context
import android.os.Environment
import android.text.format.DateFormat
import android.util.Log
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import com.alamin5g.pdf.PDFView
import com.google.firebase.FirebaseApp
import com.google.firebase.appcheck.FirebaseAppCheck
import com.google.firebase.appcheck.playintegrity.PlayIntegrityAppCheckProviderFactory
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageMetadata
import dagger.hilt.android.HiltAndroidApp
import java.io.File
import java.io.FileOutputStream
import java.util.Calendar
import java.util.HashMap
import java.util.Locale

@HiltAndroidApp
class MyApplication : Application() {

    override fun onCreate() {
        super.onCreate()
        FirebaseApp.initializeApp(this)

        val appCheckProviderFactory = if (BuildConfig.DEBUG) {
            com.google.firebase.appcheck.debug.DebugAppCheckProviderFactory.getInstance()
        } else {
            PlayIntegrityAppCheckProviderFactory.getInstance()
        }
        
        FirebaseAppCheck.getInstance()
            .installAppCheckProviderFactory(appCheckProviderFactory)
    }

    companion object {
        const val TAG_DOWN = "DOWN_BOOK"

        @JvmStatic
        fun formatTimeStamp(timestamp: Long): String {
            val cal = Calendar.getInstance(Locale.ENGLISH)
            cal.timeInMillis = timestamp
            return DateFormat.format("dd/MM/yyyy", cal).toString()
        }

        @JvmStatic
        fun deleteBook(context: Context, bookId: String, bookUrl: String, bookTitle: String) {
            val TAG = "DELETE_BOOK"
            Log.d(TAG, "deleteBook: ")

            val progressDialog = ProgressDialog(context)
            progressDialog.setTitle("Please wait")
            progressDialog.setMessage("Deleting $bookTitle...")
            progressDialog.show()

            val storageReference = FirebaseStorage.getInstance().getReferenceFromUrl(bookUrl)
            storageReference.delete()
                .addOnSuccessListener {
                    val reference = FirebaseDatabase.getInstance().getReference("Books")
                    reference.child(bookId)
                        .removeValue()
                        .addOnSuccessListener {
                            progressDialog.dismiss()
                            Toast.makeText(context, "Book delete successfully", Toast.LENGTH_SHORT).show()
                        }
                        .addOnFailureListener { e ->
                            progressDialog.dismiss()
                            Toast.makeText(context, "" + e.message, Toast.LENGTH_SHORT).show()
                        }
                }
                .addOnFailureListener { e ->
                    progressDialog.dismiss()
                    Toast.makeText(context, "" + e.message, Toast.LENGTH_SHORT).show()
                }
        }

        @JvmStatic
        fun loadPdfSize(pdfUrl: String, pdfTitle: String, sizeTv: TextView) {
            val TAG = "BOOK_SIZE_TAG"

            val ref = FirebaseStorage.getInstance().getReferenceFromUrl(pdfUrl)
            ref.metadata
                .addOnSuccessListener { storageMetadata ->
                    Log.d(TAG, "onSuccess: $pdfTitle")

                    val bytes = storageMetadata.sizeBytes.toDouble()
                    val kb = bytes / 1024
                    val mb = kb / 1024

                    if (mb >= 1) {
                        sizeTv.text = String.format("%.2f MB", mb)
                    } else if (kb >= 1) {
                        sizeTv.text = String.format("%.2f KB", kb)
                    } else {
                        sizeTv.text = String.format("%.2f bytes", bytes)
                    }
                }
                .addOnFailureListener { e ->
                    Log.d(TAG, "onFailure: " + e.message)
                }
        }

        @JvmStatic
        fun loadBannerPdf(
            pdfUrl: String,
            pdfTitle: String,
            pdfView: PDFView,
            progressBar: ProgressBar,
            pageTv: TextView?
        ) {
            val TAG = "LOAD_BANNER_PDF"
            Log.d(TAG, "loadBannerPdf: url la: $pdfUrl")

            val ref = FirebaseStorage.getInstance().getReferenceFromUrl(pdfUrl)
            ref.getBytes(Constants.MAX_BYTES_PDF)
                .addOnSuccessListener { bytes ->
                    Log.d(TAG, "onSuccess: load$pdfTitle")

                    pdfView.fromBytes(bytes)
                        .pages(0) // show ra trang dau tien cua pdf
                        .spacing(0)
                        .swipeHorizontal(false)
                        .enableSwipe(false)
                        .onError { t: Throwable ->
                            progressBar.visibility = View.INVISIBLE
                            Log.d(TAG, "onError: " + t.message)
                        }
                        // .onPageError { page, t -> ... } // Not supported in this version
                        .onLoad { nbPages: Int ->
                            progressBar.visibility = View.INVISIBLE
                            pageTv?.text = "$nbPages"
                        }
                        .load()
                }
                .addOnFailureListener { e ->
                    Log.d(TAG, "onFailure: " + e.message)
                }
        }

        @JvmStatic
        fun loadCategory(categoryId: String, categoryTv: TextView) {
            val ref = FirebaseDatabase.getInstance().getReference("Categories")
            ref.child(categoryId)
                .addListenerForSingleValueEvent(object : ValueEventListener {
                    override fun onDataChange(snapshot: DataSnapshot) {
                        val category = "${snapshot.child("category").value}"
                        categoryTv.text = category
                    }

                    override fun onCancelled(error: DatabaseError) {
                    }
                })
        }

        @JvmStatic
        fun incrementBookCount(bookId: String) {
            val ref = FirebaseDatabase.getInstance().getReference("Books")
            ref.child(bookId)
                .addListenerForSingleValueEvent(object : ValueEventListener {
                    override fun onDataChange(snapshot: DataSnapshot) {
                        var viewCount = "${snapshot.child("viewCount").value}"
                        if (viewCount == "" || viewCount == "null") {
                            viewCount = "0"
                        }

                        val newViewCount = viewCount.toLong() + 1
                        val hashMap = HashMap<String, Any>()
                        hashMap["viewCount"] = newViewCount

                        val reference = FirebaseDatabase.getInstance().getReference("Books")
                        reference.child(bookId).updateChildren(hashMap)
                    }

                    override fun onCancelled(error: DatabaseError) {
                    }
                })
        }

        @JvmStatic
        fun downloadBook(context: Context, bookId: String, bookTitle: String, bookUrl: String) {
            val namepdf = "$bookTitle.pdf"
            Log.d(TAG_DOWN, "downloadBook: $namepdf")

            val progressDialog = ProgressDialog(context)
            progressDialog.setTitle("Please wait")
            progressDialog.setMessage("Downloading $namepdf")
            progressDialog.setCanceledOnTouchOutside(false)
            progressDialog.show()

            val storageReference = FirebaseStorage.getInstance().getReferenceFromUrl(bookUrl)
            storageReference.getBytes(Constants.MAX_BYTES_PDF)
                .addOnSuccessListener { bytes ->
                    Log.d(TAG_DOWN, "onSuccess: Saving book")
                    saveDownloadBook(context, progressDialog, bytes, namepdf, bookId)
                }
                .addOnFailureListener { e ->
                    progressDialog.dismiss()
                    Toast.makeText(context, "Failed to download " + e.message, Toast.LENGTH_SHORT).show()
                }
        }

        @JvmStatic
        private fun saveDownloadBook(
            context: Context,
            progressDialog: ProgressDialog,
            bytes: ByteArray,
            namepdf: String,
            bookId: String
        ) {
            Log.d(TAG_DOWN, "saveDownloadBook: Saving download book")
            try {
                val downloadFolder =
                    Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
                downloadFolder.mkdirs()

                val filePath = downloadFolder.path + "/" + namepdf
                val out = FileOutputStream(filePath)
                out.write(bytes)
                out.close()

                Toast.makeText(context, "Saved to Download Folder", Toast.LENGTH_SHORT).show()
                progressDialog.dismiss()

                incrementBookDownloadCount(bookId)

            } catch (e: Exception) {
                Log.d(TAG_DOWN, "saveDownloadBook: Failed saving to Download Folder: " + e.message)
                Toast.makeText(context, "Failed saving to Download: " + e.message, Toast.LENGTH_SHORT).show()
            }
        }

        @JvmStatic
        private fun incrementBookDownloadCount(bookId: String) {
            val ref = FirebaseDatabase.getInstance().getReference("Books")
            ref.child(bookId)
                .addListenerForSingleValueEvent(object : ValueEventListener {
                    override fun onDataChange(snapshot: DataSnapshot) {
                        var downloadCount = "${snapshot.child("downloadCount").value}"
                        Log.d(TAG_DOWN, "onDataChange: DL count: $downloadCount")
                        if (downloadCount == "" || downloadCount == "null") {
                            downloadCount = "0"
                        }

                        val newDownloadCount = downloadCount.toLong() + 1
                        val hashMap = HashMap<String, Any>()
                        hashMap["downloadCount"] = newDownloadCount

                        val reference = FirebaseDatabase.getInstance().getReference("Books")
                        reference.child(bookId)
                            .updateChildren(hashMap)
                            .addOnSuccessListener {
                                Log.d(TAG_DOWN, "onSuccess: incrementBookDownload ")
                            }
                            .addOnFailureListener { e ->
                                Log.d(TAG_DOWN, "onFailure: failed update " + e.message)
                            }
                    }

                    override fun onCancelled(error: DatabaseError) {
                    }
                })
        }

        @JvmStatic
        fun addToFavorite(context: Context, bookId: String) {
            val firebaseAuth = FirebaseAuth.getInstance()
            if (firebaseAuth.currentUser == null) {
                Toast.makeText(context, "You are not logged in", Toast.LENGTH_SHORT).show()
            } else {
                val timestamp = System.currentTimeMillis()

                val hashMap = HashMap<String, Any>()
                hashMap["bookId"] = bookId
                hashMap["timestamp"] = "$timestamp"

                val ref = FirebaseDatabase.getInstance().getReference("Users")
                ref.child(firebaseAuth.uid!!).child("Favorites").child(bookId)
                    .setValue(hashMap)
                    .addOnSuccessListener {
                        Toast.makeText(context, "Added to your favorite list", Toast.LENGTH_SHORT).show()
                    }
                    .addOnFailureListener { e ->
                        Toast.makeText(context, "Failed add to your favorite list " + e.message, Toast.LENGTH_SHORT).show()
                    }
            }
        }

        @JvmStatic
        fun removeFromFavorite(context: Context, bookId: String) {
            val firebaseAuth = FirebaseAuth.getInstance()
            if (firebaseAuth.currentUser == null) {
                Toast.makeText(context, "You are not logged in", Toast.LENGTH_SHORT).show()
            } else {
                val ref = FirebaseDatabase.getInstance().getReference("Users")
                ref.child(firebaseAuth.uid!!).child("Favorites").child(bookId)
                    .removeValue()
                    .addOnSuccessListener {
                        Toast.makeText(context, "Removed from your favorite list", Toast.LENGTH_SHORT).show()
                    }
                    .addOnFailureListener { e ->
                        Toast.makeText(context, "Failed to remove from your favorite list " + e.message, Toast.LENGTH_SHORT).show()
                    }
            }
        }
    }
}
